GeoIp location

ptesavol opened the pull request:
* Add geoip-location package that can map an ip address to latitude and longitude using MaxMind's GeoLite City database

* If "geoIpDatabaseFolder" config variable is given to DhtNode, use geoip-location package server-side in DhtNode's ConnectivityRequest handler to add latitude and longitude to each ConnectivityReply

* Add function to cdn-location package for mapping coordinates to a local region number
 
* In DHTNode client-side, calculate local region number based on the latitude and longitude contained in ConnectivityReply. If ConnectivityReply does not contain latitude and longitude, fallback to using cdn-location.

* Take geoip-location server into use in Broker

ptesavol committed:
cdn-location-without-data-generation

ptesavol committed:
x# Please enter a commit message to explain why this merge is necessary,

ptesavol committed:
fix fetch timeouts

ptesavol committed:
move cdn-location higher in package.json, decrease fetch timeout to 2s

ptesavol committed:
changes according to review comments

ptesavol committed:
more review fixes

ptesavol committed:
add data geration scripts and documentation

ptesavol committed:
merged with main

ptesavol committed:
x# Please enter a commit message to explain why this merge is necessary,

ptesavol committed:
add airport coordinates to airportCodeToRegionNumber.ts

ptesavol committed:
udpate getLocalRegion.ts to handle the new airport data with coordinates

ptesavol committed:
add functionality for getting airport code and region by coordinates

ptesavol committed:
add package-lock

ptesavol committed:
add geoip packages

ptesavol committed:
add latitude and longitude to ConnectivityResponse

ptesavol committed:
make generatePeerDescriptorCallBack async

ptesavol committed:
make DhtNode get local region by coordinates

ptesavol committed:
working version with node-geolite2-redist

ptesavol committed:
own downloader in dht/helpers

ptesavol committed:
move geoip-location to separate package, fix browser tests

ptesavol committed:
merge with main

ptesavol committed:
merged with main

ptesavol committed:
fix ci

juslesan commented on packages/dht/src/dht/DhtNode.ts:
```diff
@@ -422,11 +417,22 @@ export class DhtNode extends EventEmitter<Events> implements ITransport {
+            } else {
+                // as a fallback get the region from the CDN
+                // and if it's not available, use a random region
+                region = await getLocalRegion()
```
This function does not need to be async if the cdn-region is queried during start "just in case". As the `generatePeerDescriptorCallBack` is vsllrf multiple times if autocertification is enabled it could be reasonable to do so

juslesan replied:
Or just cache the result

ptesavol replied:
Actually this function not being async has been very problematic also in the past (had to invent all kinds of workarounds), I think it is good that we finally got rid of this problem and need no workarounds ever again.

juslesan replied:
Probably still a good idea to cache the results of the cdn-location query.

ptesavol replied:
added getLocalRegionWithCache() function to cdn-location

ptesavol replied:
make cacheAge a optional parameter with default constant in cdnlocation

juslesan commented on packages/geoip-location/tsconfig.jest.json:
```diff
@@ -0,0 +1,17 @@
+    "references": [
+        { "path": "../utils/tsconfig.node.json" },
+        { "path": "../test-utils/tsconfig.node.json" },
+        { "path": "../proto-rpc/tsconfig.node.json" },
```
remove unnecessary references

ptesavol replied:
done

juslesan commented on packages/geoip-location/src/downloadGeoIpDatabase.ts:
```diff
@@ -0,0 +1,156 @@
+import NodePath from 'path'     // use NodePath to avoid conflict with other 'path' symbols
+import { CityResponse, Reader } from 'mmdb-lib'
+
+const GEOIP_MIRROR_URL = 'https://raw.githubusercontent.com/GitSquared/node-geolite2-redist/master/redist/'
```
Would it make sense to have a more static version being downloaded than the master branch? Just in case...

ptesavol replied:
This package seems to keep the db up-to-date and EULA-compliant by committing a new db to the master branch twice a week, so downloading from the master means that we get the same database as the node-geolite-redist package users. If the db in the master branch is sometimes broken, our own geoip package keeps using the old db it alredy has.

ptesavol replied:
In order to alleviate security concerns, forced the x chmod flag to be off on all downloaded files.

juslesan commented on packages/geoip-location/src/downloadGeoIpDatabase.ts:
```diff
@@ -0,0 +1,156 @@
+
+    try {
+        // move the extracted file to the correct location
+        fs.renameSync(dbPath + '.download/' + DB_NAME + DB_SUFFIX, dbPath + DB_NAME + DB_SUFFIX)
```
Make the correct path into a variable

juslesan commented on packages/geoip-location/src/downloadGeoIpDatabase.ts:
```diff
@@ -0,0 +1,156 @@
+    if (isDbValid === false) {
+        await downloadNewDb(GEOIP_MIRROR_URL + DB_NAME + TAR_SUFFFIX, dbFolder, remoteHash, abortSignal)
+        // return new reader if db was downloaded
+        return new Reader<CityResponse>(fs.readFileSync(dbFolder + DB_NAME + DB_SUFFIX))
```
make file path into a variable

juslesan commented on packages/dht/src/connection/websocket/WebsocketServerConnection.ts:
```diff
@@ -26,14 +26,19 @@ export class WebsocketServerConnection extends EventEmitter<ConnectionEvents> im
         this.socket = socket
     }
+    
+    // use a getter to make it possible to mock the value in tests
```
The value is not mocked in any tests so could just revert the change

ptesavol replied:
Yes it is mocked, for example here: https://github.com/streamr-dev/network/blob/7284f0195dd7380fe8095c0b8026d0e3a03f2ae7/packages/dht/test/end-to-end/GeoIpLayer0.test.ts#L34

juslesan commented on packages/cdn-location/data-generation/final-data/airportCodeToRegionNumber.ts:
```diff
@@ -1,345 +1,345 @@
-    DCA: 34100,
-    PHL: 34200,
-    NYC: 34300,
+export const airportCodeToRegionNumber: Record<string, [regionNumber: number, latitude: number, longitude: number]> = {
```
Could rename the function since it now also returns coordinates.

juslesan replied:
Could also return the value in a map instead of an array to make the code that uses the function more understandable.

ptesavol replied:
This is actually generated data (not a function), and there are a lot of code that depends on the structure of this data (the data gets sorted etc). The file is large already, and converting it into objects would almost double its size. I think it is better not to touch this.

ptesavol replied:
renamed airportCodeToRegionNumber to airportCodeToRegion

juslesan commented on packages/geoip-location/src/downloadGeoIpDatabase.ts:
```diff
@@ -0,0 +1,156 @@
+const DB_SUFFIX = '.mmdb'
+const HASH_SUFFIX = '.mmdb.sha384'
+
+const downloadNewDb = async (url: string, dbPath: string, remoteHash: string, 
```
Wondering what happens if permissions do not allow for writing or reading in the directory

ptesavol replied:
It throws, tested here: https://github.com/streamr-dev/network/blob/7284f0195dd7380fe8095c0b8026d0e3a03f2ae7/packages/geoip-location/test/unit/downloadGeoIpDatabase.test.ts#L19

juslesan reviewed (commented):
Some comments

ptesavol committed:
rename geoIpDatabasePath to geoIpDatabaseFolder for clarity

ptesavol committed:
remove extra references from tsconfig

ptesavol committed:
precompute paths to variables in the beginning of functions, set permissions of downloaded files

ptesavol committed:
update package lock

ptesavol committed:
rename airportCodeToRegionNumber to airportCodeToRegion

ptesavol committed:
add getLocalRegionWithCache()

ptesavol committed:
merge with main

ptesavol committed:
update README to include airportCodeToRegionNumber -> airportCodeToRegion change

ptesavol committed:
regenerated package lock

ptesavol committed:
take package-lock from main

ptesavol committed:
fix tests in ci

ptesavol committed:
add geoip-location references to tsconfigs

ptesavol committed:
trying to fix build

juslesan committed:
fix build

ptesavol committed:
add local caching of db files into tests to make them work on slow internet connections

ptesavol committed:
eslint

ptesavol committed:
increase test times in test/unit/downloadGeoIpDatabase.test.ts

ptesavol committed:
increase timeouts in test/unit/GeoIpLocator.test.ts

ptesavol committed:
increase timeout on GeoIpLocator-intervals.test.ts

ptesavol committed:
decrease timeout on GeoIpLocator-intervals.test.ts

ptesavol committed:
wait dbs to be downloaded in test/unit/GeoIpLocator-intervals.test.ts

ptesavol committed:
add caching on filesystem to TestServer

ptesavol committed:
merge with main

ptesavol committed:
package-lock from main

ptesavol committed:
version update

ptesavol committed:
resolve race condition

ptesavol committed:
fix dependencies

ptesavol committed:
concurrency fixes

ptesavol committed:
remove too strict memoty test

Commit:
Take geoip-location server into use in Broker (#2503)

## Summary

* Add geoIpDatabaseFolder config variable to Client configs and pass it
to DhtNode
* Set the geoIpDatabaseFolder config variable to
'~/.streamr/geoipdatabases' in Broker, which turns on geoip server
functionality in Broker, but leaves it off in other cases

Co-authored-by: Petri Savolainen <petri.savolainen@spaceify.net>

juslesan commented:
Should add the geoip-location package to the `release.sh` file

juslesan commented on packages/dht/src/connection/websocket/WebsocketConnector.ts:
```diff
@@ -77,6 +80,7 @@ export class WebsocketConnector {
             maxMessageSize: config.maxMessageSize,
             enableTls: config.serverEnableTls
         }) : undefined
+
```
```suggestion
```

ptesavol replied:
done

juslesan commented on packages/dht/src/connection/websocket/WebsocketConnector.ts:
```diff
@@ -159,12 +163,22 @@ export class WebsocketConnector {
                     } else {
                         logger.trace('incoming Websocket connection before localPeerDescriptor was set, closing connection')
-                        connection.close(false).catch(() => {})
+                        connection.close(false).catch(() => { })
```
```suggestion
                        connection.close(false).catch(() => {})
```

juslesan commented on packages/dht/src/connection/websocket/WebsocketConnector.ts:
```diff
@@ -159,12 +163,22 @@ export class WebsocketConnector {
             const port = await this.websocketServer.start()
             this.selectedPort = port
+
+            if (this.config.geoIpDatabaseFolder) {
```
Could encapsulate the geoip-location related logic to a facade module.

juslesan replied:
Perhaps it would make sense to start the geoip-location before setting the server listener?

ptesavol replied:
Changed the starting order, I don't think there is a need for a facade for such a simple component as this.

juslesan commented on packages/dht/src/connection/websocket/WebsocketServerConnection.ts:
```diff
@@ -26,14 +26,19 @@ export class WebsocketServerConnection extends EventEmitter<ConnectionEvents> im
     }
+    
+    // use a getter to make it possible to mock the value in tests
+    public get remoteIpAddress(): string {
```
```suggestion
    public getRemoteIpAddress(): string {
```

ptesavol replied:
The use of an actual js getter is intentional, the user of WebSocketServerConnection does not need to know this field has been changed into a getter.

ptesavol replied:
change to java-style getter function

juslesan commented on packages/geoip-location/src/GeoIpLocator.ts:
```diff
@@ -0,0 +1,113 @@
+    // By default, check the database every 30 days
+    // If the check fails, retry after 24 hours
+
+    constructor(geoIpDatabaseFolder: string, dbCheckInterval: number = 30 * 24 * 60 * 60 * 1000,
```
Make this default checkIntervals to contants

juslesan commented on packages/geoip-location/test/unit/GeoIpLocator-intervals.test.ts:
```diff
@@ -0,0 +1,107 @@
+
+    beforeAll(async () => {
+        testServer = new TestServer()
+        await testServer.start(31991)
```
```suggestion
        await testServer.start(serverPort)
```

juslesan commented on packages/geoip-location/test/unit/tarHelper.test.ts:
```diff
@@ -0,0 +1,102 @@
+    sPort++
+    return port
+}
+describe('tarHelper', () => {
```
Start and stop server is beforeEach and afterEach

ptesavol replied:
Putting server start and stop in tests themselves is intentional. The servers are unique, they run on different ports and some of them get aborted in the middle. I tried using beforeEach() and afterEach(), which resulted in the servers of wrong test tests cases getting closed.

ptesavol replied:
Try beforeEach/afterEach with same port number for all test cases

juslesan commented on packages/geoip-location/test/unit/GeoIpLocator-no-network-at-monthly.test.ts:
```diff
@@ -0,0 +1,53 @@
+
+    beforeAll(async () => {
+        testServer = new TestServer()
+        await testServer.start(31990)
```
add serverPort constant, use it on line 23 as well

juslesan commented on packages/geoip-location/test/unit/GeoIpLocator.test.ts:
```diff
@@ -0,0 +1,125 @@
+
+    beforeAll(async () => {
+        testServer = new TestServer()
+        await testServer.start(31992)
```
add serverPort constant, use it on line 31 as well

juslesan reviewed (changes requested):
Should we remove the geoip-location packages from the browser bundle in the client?

ptesavol committed:
respond to review comments

ptesavol committed:
add debug logging

ptesavol committed:
merge with main

ptesavol committed:
move TestServer in tarHelper.test.ts to use beforeEach/afterEach

ptesavol committed:
move to using java-style getter getRemoteIpAddress() in ServerWebSocketConnection

ptesavol committed:
use default cache age in getLocalRegionWithCache()

ptesavol committed:
Merge branch 'main' into geoip-location

ptesavol committed:
change port of GeoIpConnectivityChecking.test.ts

ptesavol committed:
add more time for express to start up

ptesavol committed:
merged

juslesan committed:
remove unnecessary undefined types

juslesan committed:
refactor & reformat

juslesan committed:
reduce test tiemouts

juslesan committed:
move pipeline

juslesan committed:
eslint

juslesan committed:
Merge branch 'main' into geoip-location

juslesan commented on packages/geoip-location/src/GeoIpLocator.ts:
```diff
@@ -0,0 +1,109 @@
+        // this might happen despite the ts typings because the ip address 
+        // comes from the ws server socket and is not under our control
+
+        if (!(ip as unknown)) {
```
Is there any case where this would happen in the real world?

ptesavol replied:
It did happen while testing, and that's why I added the check. 

In all, this is a timer-run system that is run very rarely, so we should make it as difficult to crash as possible, as crashes have a high probability of staying unnoticed ("It has been running fine for 20 days, so probably it will keep that way")

juslesan committed:
fix mock typing

juslesan committed:
reformat

juslesan committed:
protocol version 1.1

juslesan committed:
reformat

juslesan committed:
refactor cache

juslesan committed:
move mock restore

juslesan committed:
Merge branch 'main' into geoip-location

juslesan committed:
Merge branch 'main' into geoip-location

juslesan committed:
undefined check for weird getRemoteIpAddress cases

juslesan committed:
fix test

juslesan committed:
falsy

juslesan committed:
Merge branch 'main' into geoip-location

juslesan committed:
Changelog
